blog/2023-08-12-why-mlops-sucks/index.md

```diff
@@ -0,0 +1,781 @@
+                sigma = self.scheduler.sigmas[i]
+                latent_model_input = latent_model_input / ((sigma**2 + 1) ** 0.5)
+
+                # predict the noise residual
+                inputs = [
+                    latent_model_input,
+                    torch.tensor([t]).to(self.device),
+                    text_embeddings,
+                ]
+                noise_pred = self.unet(inputs, timing=True)
+                noise_pred = torch.reshape(noise_pred[0], (batch_size*2, 4, 64, 64))
+
+                noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
+                noise_pred = noise_pred_uncond + guidance_scale * (
+                    noise_pred_text - noise_pred_uncond
+                )
+                # compute the previous noisy sample x_t -> x_t-1
+                latents = self.scheduler.step(noise_pred.cuda(), t, latents)["prev_sample"]
+
+            # scale and decode the image latents with vae
+            latents = 1 / 0.18215 * latents
+            image = self.vae.decode(latents).sample
+        return image
+
+model = TrtDiffusionModel()
+image = model.predict(
+    prompts="Iron man laughing, real photoshoot",
+    num_inference_steps=25,
+    height=512,
+    width=512,
+    max_seq_length=64,
+)
+image = (image / 2 + 0.5).clamp(0, 1)
+image = image.detach().cpu().permute(0, 2, 3, 1).numpy()
+images = (image * 255).round().astype("uint8")
+pil_images = [Image.fromarray(image) for image in images]
+pil_images[0].save("image_generated.png")
+```
+
+It seems that when the above script is ran, it works but the generations look something like these:
+
+<!-- ![black image](black_image.png)
+
+![noise image](noise_image.png) -->
+
+black image         |  noise image
+:-------------------------:|:-------------------------:
+![black image](black_image.png) |  ![noise image](noise_image.png)
+
+which shows that something’s going wrong here, and changing to different tensor shapes (defined above) also doesn’t help fix noisy or no generations. It’s a big blocker for me to proceed with stable-diffusion 2.1 converted into TensorRT engine format for inference. Though it was possible for few other variants of Stable Diffusion coming from [AUTOMATIC1111/stable-diffusion-webui](https://github.com/AUTOMATIC1111/stable-diffusion-webui)
+
+With some time searching about this, found [this](https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/5503#issuecomment-1341495770) relevant issue with Stable Diffusion giving out black images. It mentions two ways i.e:
+
+- Do conversions of this model in > fp16 - which I did but it made no difference and I was still getting noise or black images only.
+- Using xformers but currently for onnx conversion we need pytorch nightly since there’s scaled_dot_product_attention operator which isn’t supported to export in normal pytorch but only through nightly - refer to [this](https://github.com/pytorch/pytorch/issues/97262) issue.
+
+## Conclusion
```

Comment:
would be nice to link to a repo with code (maybe one folder per experiment).
That way you could reduce the amount of code in this post

Reply:
I agree that it's making the post a bit lengthy but if I put the code under a repository then I'll be just linking the code, and since this blog contains lots of code - reader would have to switch context a lot from repo <-> post. Also a reading time calculator shows that it will take ~18mins for this post (but yes we know the time to read will be longer than this since it's treating code as normal words).